File: character-predictor/src/main/Predictor.scala
package main

import com.logitech.gaming.LogiLED

/**
  * Created by blueur on 19/11/16.
  */
class Predictor {
  private val probs = new ProbabilitiesDatabase(10)
  private val alphabet = "abcdefghijklmnopqrstuvwxyz -'"
  private val n = 10

  probs.initialize("big.txt")

  def computeProbability(prev: String, char: Char): Double = {
    probs.getNgramProbabilities(prev + char)/probs.getNgramProbabilities(prev)
  }

  def getNextChar(text: String): List[(Char, Double)] = {
    /* Not enough chars to do a prediction */
    if (text.length < 1) {
      return List(('\0', 0.0))
    }

    val last = text.takeRight(n).toLowerCase()

    alphabet.map(c => (c, computeProbability(last, c)))
            .sortWith((a, b) => a._2 > b._2)
            .toList
  }
}

object Main {
  def main(argv: Array[String]): Unit = {
    val keyboard = new KeyboardMessageDisplay()
    val predictor = new Predictor()

    var last = ' '
    var read = ' '
    var text = ""

    LogiLED.LogiLedInit()
    LogiLED.LogiLedSetLighting(0, 0, 0)

    while (read != '\r' && read != '\n') {
      read = io.StdIn.readChar()
      text += read

      var letters = predictor.getNextChar(text)
      var key = letters.head._1

      keyboard.ShowLetter(key, 100, 0, 0)
      keyboard.ShowLetter(last, 100, 0, 0)

      last = key

      letters.foreach(cp => print("%s (%.2f)".format(cp._1, cp._2)))
      println()
    }
  }
}

File: KeyboardPredictor/src/ch/lauzhack/Predictor.java
package ch.lauzhack;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.logitech.gaming.LogiLED;

/**
 * Created by Loic on 19.11.2016.
 */
public class Predictor {
    private ProbabilitiesDatabase probs;
    private String alphabet = "abcdefghijklmnopqrstuvwxyz -'";
    private int n;

    public Predictor(int n) {
        this.probs = new ProbabilitiesDatabase(n);
        this.n = n;
        KeyboardMessageDisplay kb = new KeyboardMessageDisplay();
        kb.showHeart(true);
        probs.loadTextFile("big.txt");
        probs.loadTextFile("words2.txt");
        kb.showHeart(false);
//		LogiLED.LogiLedSetLighting(0, 0, 0);
    }

    public boolean isValidChar(char c) {
        return alphabet.contains("" + c);
    }

    public double computeProbability(String prev, char next) {
        double denom = probs.getNgramProbabilities(prev);
        if (denom <= 0) {
            return 0;
        }
        return probs.getNgramProbabilities(prev + next)/denom;
    }

    public List<CharProbPair> getNextChar(String text) {
        List<CharProbPair> letters = new ArrayList<>();

        String last = text;

        letters.clear();
        for (char c : alphabet.toCharArray()) {
            letters.add(new CharProbPair(c, computeProbability(last, c)));
        }
        Collections.sort(letters);

        // No prediction ? Only show space
        if (letters.get(0).getProbability() == 0) {
            letters.clear();
            letters.add(new CharProbPair(' ', 1.0));
        }

        return letters;
    }
}
